Add ModUtils class for validating mod usages

smoogipoo opened the pull request:
Useful for both osu! and server-side validation.

smoogipoo committed:
Add ModUtils class for validating mod usages

bdach commented on osu.Game/Utils/ModUtils.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <returns>Whether all <see cref="Mod"/>s in the combination are compatible with each-other.</returns>
+        public static bool CheckCompatibleSet(IEnumerable<Mod> combination)
+        {
+            var incompatibleTypes = new HashSet<Type>();
```
This hashset is added to, but never queried. (Weird that there's no inspection on this?)

In fact I'm not even sure I understand what the two hashsets are supposed to be tracking...

peppy replied:
I was looking at this method with the intention of combining (a portion of?) it with my follow-up function in #11642, but am also unsure how to read this code. Like I don't understand the early return or what `incomingTypes` is tracking. Is there a reason you didn't write this in a similar way to what I have in my PR?

smoogipoo replied:
Re: never queried: It's supposed to be querying it in the second block, not sure how it passes the tests.

It's actually doing about the same thing your code is, just in a different way. Wouldn't mind replacing it. Though I think your code should read `type.IsInstanceOfType(m)`?

peppy replied:
![20210202 140924 (JetBrains Rider)](https://user-images.githubusercontent.com/191335/106555123-42634580-6560-11eb-9a20-b8fc8c750f00.png)

That's how it reads on the current version, did you mean something else?

bdach commented on osu.Game.Tests/Mods/ModUtilsTest.cs:
```diff
@@ -0,0 +1,64 @@
+        public void TestModIsCompatibleByItself()
+        {
+            var mod = new Mock<Mod>();
+            Assert.That(ModUtils.CheckCompatibleSet(new[] { mod.Object }));
```
huh, TIL that it's not required to pass `Is.True`

smoogipoo committed:
Add failing tests

smoogipoo committed:
Fix incorrect implementation

smoogipoo committed:
Simplify implementation of CheckCompatibleSet

smoogipoo committed:
Fix incorrect handling of multi-mod incompatibilities

smoogipoo committed:
Make more tests use the custom mod classes

For safety purposes... In implementing the previous tests, I found that
using mod.Object.GetType() can lead to bad assertions since the same
ModProxy class is used for all mocked classes.

smoogipoo committed:
Prevent multiple enumeration

smoogipoo commented:
Should be ready to go now. Also made some changes in the dependant PR.

smoogipoo committed:
Add codeanalysis attribute
